cgp/local_search/evolution_strategies.py

```diff
@@ -0,0 +1,209 @@
+            distribution. Defaults to (3 + log(dim)) / (5.0 *
+            sqrt(dim)), where dim is the dimension of the search space.
+        population_size : int, optional
+            Number of samples evaluated per generation. Defaults to 4
+            + floor(3 * log(dim))), where dim is the dimension of the
+            search space..
+        max_generations : int, optional
+            Maximal number of generations. Defaults to 10.
+        min_sigma : float, optional
+            Minimal value for standard deviation of search
+            distribution. Defaults to 1e-9.
+        fitness_shaping : bool, optional
+            Whether to use fitness shaping. Defaults to True.
+        mirrored_sampling : bool, optional
+            Whether to use mirrored sampling. WARNING: Doubles the number
+            of samples per generation. Defaults to True.
+
+        """
+        self.objective = objective
+        self.seed = seed
+        self.learning_rate_mu = learning_rate_mu
+        self.learning_rate_sigma = learning_rate_sigma
+        self.population_size = population_size
+        self.max_generations = max_generations
+        self.min_sigma = min_sigma
+        self.fitness_shaping = fitness_shaping
+        self.mirrored_sampling = mirrored_sampling
+
+        self.sigma: Dict[int, Dict[str, np.ndarray[float]]] = collections.defaultdict(dict)
+
+    def __call__(self, ind: "IndividualBase") -> None:
+
+        rng = np.random.RandomState(self.seed)
+
+        mu: np.ndarray[float]
+        params_names: List[str]
+        mu, params_names = ind.parameters_to_numpy_array(only_active_nodes=True)
+
+        sigma: np.ndarray[float] = self._load_sigma(ind, mu, params_names)
+
+        if len(mu) > 0:
+            if self.learning_rate_sigma is None:
+                learning_rate_sigma = (3 + math.log(len(mu))) / (5.0 * math.sqrt(len(mu)))
+            else:
+                learning_rate_sigma = self.learning_rate_sigma
+
+            if self.population_size is None:
+                population_size = 4 + int(math.floor(3 * math.log(len(mu))))
+            else:
+                population_size = self.population_size
+
+            obj: Callable[[np.ndarray[float]], float] = functools.partial(
+                self._objective_wrapper, ind=ind, params_names=params_names
+            )
+
+            for generation in range(self.max_generations):
+
+                s: np.ndarray[float]
+                z: np.ndarray[float]
+                s, z = self._sample_s_and_z(mu, sigma, population_size, rng)
+
+                fitness: np.ndarray[float] = np.fromiter(map(obj, z), np.float)
+
+                s, z, utility = self._determine_utility(s, z, fitness)
+
+                mu, sigma = self._update_parameters_of_search_distribution(
+                    s, utility, learning_rate_sigma, mu, sigma
+                )
+
+                if np.all(sigma < self.min_sigma):
+                    break
```

Comment:
Is this to stop the local search, if it has "converged" to a narrow interval?

Reply:
yes, typically convergence of sigma is a sufficient criterion for convergence of the search. but i think this should be removed to avoid suprises. in that case always `max_generations` steps will be performed, easy to understand.